storage_v2/spanstore/factory.go

```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package spanstore
+
+// Factory defines an interface for a factory that can create implementations of different storage components.
+// Implementations are also encouraged to implement plugin.Configurable interface.
+type Factory interface {
+	// Initialize performs internal initialization of the factory, such as opening connections to the backend store.
+	// It is called after all configuration of the factory itself has been done.
+	Initialize() error
```

Comment:
I recommend comparing this with OTEL collector component lifecycle methods. Eg would you need to accept telemetry objects here, or would it be passed to the constructor function? Do we need Close if we're adding lifecycle methods?

Also, different storage factories (eg dependencies storage) would need the same lifestyle interface so it should be pulled to the higher pkg.

Reply:
We would pass the telemetry objects to the constructor and only receive the logger and config, which is similar to the existing factory, e.g. `grpc.NewFactory(*zap.Logger, grpc.Config)`. I've looked into other OTEL collector extension implementations and they (mostly) only accept logger from the telemetry object.

I've added the flowchart of OTEL collector component's lifecycle below. Inside the `Component.Start(context.Context, component.Host)`, we will call `spanstore.NewFactory(*zap.Logger, component.Config)` and then `Factory.Initialize(context.Context)`. Once the component is shutting down through `Component.Shutdown(context.Context)`, we will call `Factory.Close(context.Context)`.

```mermaid
flowchart TD
    extension.Factory --> Factory.CreateExtension
    Factory.CreateExtension -->|returns| component.Component
    component.Component --> Component.Start
    Component.Start --> spanstore.NewFactory
    spanstore.NewFactory -->|returns| spanstore.Factory
    spanstore.Factory --> Factory.Initialize
    Factory.Initialize --> Running
    Running --> Factory.Close
    Factory.Close -->|called by| Component.Shutdown

    subgraph "OTEL Collector"
        extension.Factory
        Factory.CreateExtension
        component.Component
        Component.Start
        Component.Shutdown

        subgraph "Jaeger Storage Extension"
            spanstore.NewFactory
            spanstore.Factory
            Factory.Initialize
            Running
            Factory.Close
        end
    end
```